gdal/frmts/nitf/nitffile.c

```diff
@@ -2204,27 +2251,75 @@ static char** NITFGenericMetadataReadTREInternal(char **papszMD,
                 pszMDItemName = CPLStrdup(
                             CPLSPrintf("%s%s", pszMDPrefix, pszName));
 
-                NITFExtractMetadata( &papszTmp, pachTRE, *pnTreOffset,
-                                     nLength, pszMDItemName );
-                if (*pnMDSize + 1 >= *pnMDAlloc)
+                if (strcmp(pszType, "float_raw") == 0)
                 {
-                    *pnMDAlloc = (*pnMDAlloc * 4 / 3) + 32;
-                    papszMD = (char**)CPLRealloc(papszMD, *pnMDAlloc * sizeof(char*));
+                    if( nLength == 4 )
+                    {
+                        const size_t nBufferSize = 128;
+                        float f;
+                        memcpy(&f, pachTRE + *pnTreOffset, 4);
+                        CPL_MSBPTR32(&f);
+                        pszValue = (char*)CPLMalloc(nBufferSize);
+                        CPLsnprintf(pszValue, nBufferSize, "%f", f);
+                        papszTmp = CSLSetNameValue(papszTmp, pszMDItemName, pszValue);
+                    }
+                    else
+                    {
+                        *pbError = TRUE;
+                        CPLError( CE_Warning, CPLE_AppDefined,
+                                  "Raw float field must be 4 bytes in %s TRE ", pszTREName);
+                        break;
+                    }
                 }
-                papszMD[*pnMDSize] = papszTmp[0];
-                papszMD[(*pnMDSize) + 1] = NULL;
-                (*pnMDSize) ++;
-                papszTmp[0] = NULL;
-                CPLFree(papszTmp);
+                else if (strcmp(pszType, "bitmask") == 0)
+                {
+                    if( nLength <= 4 )
+                    {
+                        const size_t nBufferSize = 11;
+                        unsigned int nVal = 0;
+                        memcpy(&nVal, pachTRE + *pnTreOffset, nLength);
```

Comment:
I'm not smart enough to assert the correctness of this if nLength != 4. If nLength != 4 must be really supported, then I'd suggest to have a for loop on nLength to extract byte by byte and do the appropriate bit shifting to form nVal. This will be clearer / more portable

Reply:
Reverted this one for now.  nLength < 4 bytes will be needed for another TRE, so I'll take your implementation suggestion into another PR.  That TRE will also be a natural test of nLength < 4.